rustls/src/msgs/deframer.rs

```diff
@@ -431,50 +461,40 @@ impl<'a> DeframerSliceBuffer<'a> {
     }
 }
 
-impl DeframerBuffer for DeframerSliceBuffer<'_> {
-    fn advance(&mut self, _num_bytes: usize) {
-        unimplemented!("operation not supported by this type of buffer")
-    }
-
+impl FilledDeframerBuffer for DeframerSliceBuffer<'_> {
     fn filled_mut(&mut self) -> &mut [u8] {
         &mut self.buf[self.discard..]
     }
 
-    fn unfilled(&mut self) -> &mut [u8] {
-        unimplemented!("operation not supported by this type of buffer")
-    }
-
     fn filled(&self) -> &[u8] {
         &self.buf[self.discard..]
     }
 }
 
-trait DeframerBuffer {
+impl DeframerBuffer<false> for DeframerSliceBuffer<'_> {
+    fn copy(&mut self, src: &[u8], at: usize) {
+        copy_into_buffer(self.filled_mut(), src, at)
+    }
+}
+
+trait DeframerBuffer<const QUIC: bool>: FilledDeframerBuffer {
     /// Copies from the `src` buffer into this buffer at the requested index
     ///
-    /// If `quic` is true the data will be copied into the *un*filled section of the buffer
+    /// If `QUIC` is true the data will be copied into the *un*filled section of the buffer
     ///
-    /// If `quic` is false the data will be copied into the filled section of the buffer
-    fn copy(&mut self, src: &[u8], at: usize, quic: bool) {
-        let buf = if quic {
-            self.unfilled()
-        } else {
-            self.filled_mut()
-        };
-        let len = src.len();
-        let into = &mut buf[at..at + len];
-        into.copy_from_slice(src);
-        if quic {
-            self.advance(len);
-        }
-    }
+    /// If `QUIC` is false the data will be copied into the filled section of the buffer
+    fn copy(&mut self, src: &[u8], at: usize);
+}
 
-    fn advance(&mut self, num_bytes: usize);
+fn copy_into_buffer(buf: &mut [u8], src: &[u8], at: usize) {
+    let len = src.len();
+    let into = &mut buf[at..at + len];
+    into.copy_from_slice(src);
```

Comment:
Do we need the bindings for some borrowck edge cases here? If not, would be nice to just inline the oneliner.

Reply:
fixed